File: android/app/src/main/java/com/laneshadow/ui/components/atoms/RoutePolylineComponent.kt
package com.laneshadow.ui.components.atoms

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalInspectionMode
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * RoutePolylineComponent
 *
 * A higher-level route polyline component that wraps the basic RoutePolyline concept
 * with route-specific data and selection state semantics.
 *
 * This component provides a simplified API for route rendering with selection state,
 * supporting both overview routes and individual route legs.
 *
 * @param isSelected Selection state determines color styling (default: false)
 * @param coordinates List of LatLng points representing the route path
 * @param strokeWidth Width of the route line in dp for overview routes (default: 6dp)
 * @param legStrokeWidth Width of the route line in dp for route legs (default: 4dp)
 * @param onPress Optional callback when the polyline is pressed
 * @param testID Test ID for UI testing
 * @param modifier Modifier for the component container
 */
@Composable
fun RoutePolylineComponent(
    isSelected: Boolean = false,
    coordinates: List<LatLng>,
    strokeWidth: Dp = 6.dp,
    legStrokeWidth: Dp = 4.dp,
    onPress: (() -> Unit)? = null,
    testID: String? = null,
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current
    val isPreview = LocalInspectionMode.current

    // Map route colors from theme based on selection state
    val routeColor = if (isSelected) {
        theme.colors.routeSelected.default
    } else {
        theme.colors.onSurface.default.copy(alpha = 0.7f)
    }

    // Build content description for accessibility
    val contentDescription = buildRouteComponentContentDescription(
        coordinates.size,
        isSelected,
    )

    Canvas(
        modifier = modifier
            .fillMaxSize()
            .semantics {
                this.contentDescription = contentDescription
            },
    ) {
        val strokeWidthPx = strokeWidth.value * density

        // Draw route polyline
        if (coordinates.size >= 2) {
            val path = createPathFromPoints(coordinates, size.width, size.height)
            drawPath(
                path = path,
                color = routeColor,
                style = Stroke(
                    width = strokeWidthPx,
                ),
            )
        }
    }
}

/**
 * Build accessibility content description for the route component
 */
private fun buildRouteComponentContentDescription(
    pointCount: Int,
    isSelected: Boolean,
): String {
    return buildString {
        val stateText = if (isSelected) "selected" else "alternate"
        append("$stateText route. ")
        append("Contains $pointCount coordinate points.")
    }
}

/**
 * Create a Compose Path from a list of LatLng points
 *
 * This is a simplified implementation that maps geographic coordinates
 * to canvas coordinates for demonstration purposes.
 * In production with Google Maps SDK, this would be handled by the map platform.
 */
private fun createPathFromPoints(
    points: List<LatLng>,
    canvasWidth: Float,
    canvasHeight: Float,
): Path {
    val path = Path()

    if (points.isEmpty()) return path

    // Map first point
    val firstOffset = latLngToOffset(
        points[0],
        points,
        canvasWidth,
        canvasHeight,
    )
    path.moveTo(firstOffset.x, firstOffset.y)

    // Map remaining points
    for (i in 1 until points.size) {
        val offset = latLngToOffset(
            points[i],
            points,
            canvasWidth,
            canvasHeight,
        )
        path.lineTo(offset.x, offset.y)
    }

    return path
}

/**
 * Convert a LatLng point to canvas Offset
 *
 * This is a simplified projection that maps geographic coordinates
 * to canvas space for demonstration purposes.
 * Uses linear interpolation within the bounding box of all points.
 */
private fun latLngToOffset(
    point: LatLng,
    allPoints: List<LatLng>,
    canvasWidth: Float,
    canvasHeight: Float,
): Offset {
    if (allPoints.isEmpty()) {
        return Offset(canvasWidth / 2f, canvasHeight / 2f)
    }

    val minLat = allPoints.minOfOrNull { it.latitude } ?: point.latitude
    val maxLat = allPoints.maxOfOrNull { it.latitude } ?: point.latitude
    val minLon = allPoints.minOfOrNull { it.longitude } ?: point.longitude
    val maxLon = allPoints.maxOfOrNull { it.longitude } ?: point.longitude

    val latRange = maxLat - minLat
    val lonRange = maxLon - minLon

    // Add padding to prevent drawing on edges
    val padding = 20f
    val usableWidth = canvasWidth - 2 * padding
    val usableHeight = canvasHeight - 2 * padding

    val x: Float = if (lonRange > 0) {
        (padding + (point.longitude - minLon) / lonRange * usableWidth).toFloat()
    } else {
        canvasWidth / 2f
    }

    val y: Float = if (latRange > 0) {
        (padding + (1 - (point.latitude - minLat) / latRange) * usableHeight).toFloat()
    } else {
        canvasHeight / 2f
    }

    return Offset.Unspecified.copy(x = x, y = y)
}

File: ios/LaneShadow/Views/Atoms/RoutePolylineComponent.swift
import LaneShadowTheme
import SwiftUI

/// RoutePolylineComponent - Higher-level route polyline wrapper
///
/// A route-specific component that wraps RoutePolyline with simplified selection-based styling.
/// Designed for use in route cards and route selection UIs.
///
/// Following the translation matrix specification:
/// - Selected route: theme.colors.primary.default (fallback to routeSelected)
/// - Unselected route: theme.colors.onSurface.default.opacity(0.7) (fallback to routeAlternate)
/// - Stroke width: 6pt for overview, 4pt for legs
/// - Canvas height: 100pt default
public struct LSRoutePolylineComponent: View {
    // MARK: - Properties

    @Environment(\.theme) private var theme

    private let isSelected: Bool
    private let coordinates: [LSLatLng]
    private let strokeWidth: CGFloat
    private let legStrokeWidth: CGFloat
    private let height: CGFloat

    // MARK: - Initialization

    /// Creates a RoutePolylineComponent with the given route data
    /// - Parameters:
    ///   - isSelected: Whether this route is selected (default is false)
    ///   - coordinates: Array of coordinates representing the route
    ///   - strokeWidth: Width of the stroke for overview display (default is 6pt)
    ///   - legStrokeWidth: Width of the stroke for leg display (default is 4pt)
    ///   - height: Canvas height for preview (default is 100pt)
    public init(
        isSelected: Bool = false,
        coordinates: [LSLatLng],
        strokeWidth: CGFloat = 6,
        legStrokeWidth: CGFloat = 4,
        height: CGFloat = 100
    ) {
        self.isSelected = isSelected
        self.coordinates = coordinates
        self.strokeWidth = strokeWidth
        self.legStrokeWidth = legStrokeWidth
        self.height = height
    }

    // MARK: - Body

    public var body: some View {
        LSRoutePolyline(
            coordinates: coordinates,
            variant: isSelected ? .selected : .alternate,
            strokeWidth: isSelected ? strokeWidth : legStrokeWidth,
            height: height
        )
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(isSelected ? "Selected route" : "Alternate route")
        .accessibilityHint("Route polyline showing \(coordinates.count) points")
    }
}

// MARK: - Preview

#Preview("Selected Route") {
    VStack(spacing: 20) {
        LSRoutePolylineComponent(
            isSelected: true,
            coordinates: [
                LSLatLng(latitude: 37.7749, longitude: -122.4194),
                LSLatLng(latitude: 37.7750, longitude: -122.4180),
                LSLatLng(latitude: 37.7755, longitude: -122.4170),
                LSLatLng(latitude: 37.7765, longitude: -122.4160),
                LSLatLng(latitude: 37.7775, longitude: -122.4150),
            ],
            strokeWidth: 6,
            legStrokeWidth: 4,
            height: 100
        )
        .padding()

        Text("Selected Route")
            .font(.headline)
    }
    .laneShadowTheme()
}

#Preview("Unselected Route") {
    VStack(spacing: 20) {
        LSRoutePolylineComponent(
            isSelected: false,
            coordinates: [
                LSLatLng(latitude: 37.7749, longitude: -122.4194),
                LSLatLng(latitude: 37.7750, longitude: -122.4180),
                LSLatLng(latitude: 37.7755, longitude: -122.4170),
                LSLatLng(latitude: 37.7765, longitude: -122.4160),
            ],
            strokeWidth: 6,
            legStrokeWidth: 4,
            height: 100
        )
        .padding()

        Text("Unselected Route")
            .font(.headline)
    }
    .laneShadowTheme()
}

#Preview("Route Comparison") {
    VStack(spacing: 16) {
        Text("Route Options")
            .font(.headline)

        LSRoutePolylineComponent(
            isSelected: true,
            coordinates: [
                LSLatLng(latitude: 37.7749, longitude: -122.4194),
                LSLatLng(latitude: 37.7750, longitude: -122.4180),
                LSLatLng(latitude: 37.7755, longitude: -122.4170),
                LSLatLng(latitude: 37.7765, longitude: -122.4160),
                LSLatLng(latitude: 37.7775, longitude: -122.4150),
            ],
            strokeWidth: 6,
            legStrokeWidth: 4,
            height: 80
        )

        LSRoutePolylineComponent(
            isSelected: false,
            coordinates: [
                LSLatLng(latitude: 37.7749, longitude: -122.4194),
                LSLatLng(latitude: 37.7752, longitude: -122.4175),
                LSLatLng(latitude: 37.7760, longitude: -122.4170),
                LSLatLng(latitude: 37.7775, longitude: -122.4150),
            ],
            strokeWidth: 6,
            legStrokeWidth: 4,
            height: 80
        )
    }
    .padding()
    .laneShadowTheme()
}

#Preview("Empty Route") {
    VStack(spacing: 20) {
        LSRoutePolylineComponent(
            isSelected: true,
            coordinates: [],
            strokeWidth: 6,
            legStrokeWidth: 4,
            height: 100
        )
        .padding()

        Text("Empty Route (Placeholder)")
            .font(.headline)
    }
    .laneShadowTheme()
}
